Pr cmi process

ChrisLaliwala opened the pull request:
## Addresses Issue: 

## Summary/Motivation:
This PR introduces a flowsheet example for a process developed by the Critical Materials Innovation Hub (CMI) which recovers rare earth elements (REEs) from an end-of-life (EOL) rare earth permanent magnet (REPM) feedstock utilizing a series of precipitation and dissolution reactions to produce rare earth oxalates. Then, the rare earth oxalates are calcined to produce the rare earth oxides (REOs) final product. Notably, no acids are required in this process.

## Changes proposed in this PR:
- Developed a flowsheet for the CMI Process.

## Reviewer's checklist / merge requirements:
- [ ] The head branch (i.e. the "source" of the changes) is not the `main` branch on the PR author's fork
- [ ] Documentation
- [ ] Tests
- [ ] Diagnostic tests for models

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.

2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_precipitate_liquid_properties.py:
```diff
@@ -0,0 +1,276 @@
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2024 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory, et al. All rights reserved.
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license information.
+#####################################################################################################
```
Edit this.

OOAmusat replied:
This is supposed to refer to the docstring describing the file.

ChrisLaliwala replied:
Have deleted this file as there is no need for it. Using file: 'prommis.precipitate.precipitate_liquid_properties'.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_precipitate_solids_properties.py:
```diff
@@ -0,0 +1,207 @@
+"""
+Preliminary property package for West Kentucky No. 13 coal refuse.
+
+Authors: Alejandro Garciadiego
```
Edit this.

ChrisLaliwala replied:
Have deleted this file as there is no need for it. Using file: 'prommis.precipitate.precipitate_solid_properties'.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_precipitate_liquid_properties.py:
```diff
@@ -0,0 +1,276 @@
+        iscale.set_scaling_factor(self.flow_vol, 1e1)
+        iscale.set_scaling_factor(self.conc_mass_comp, 1e2)
+        iscale.set_scaling_factor(self.flow_mol_comp, 1e3)
+        iscale.set_scaling_factor(self.conc_mol_comp, 1e5)
```
These are the old scaling tools. You probably need to use the new scaling tools - see the gibbs reactor as an example:

https://github.com/IDAES/idaes-pse/blob/main/idaes/models/unit_models/gibbs_reactor.py

ChrisLaliwala replied:
This file was deleted as it was no longer needed.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_precipitate_solids_properties.py:
```diff
@@ -0,0 +1,207 @@
+            bounds=(1e-20, None),
+        )
+
+        iscale.set_scaling_factor(self.flow_mol_comp, 1e3)
```
switch to new scaling tools

ChrisLaliwala replied:
This file was deleted as it was no longer needed.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,683 @@
+
+    initialize_system(m)
+
+    if degrees_of_freedom(m) != 0:
```
for this you can just use assert `(degrees_of_freedom(m) == 0)`

ChrisLaliwala replied:
Changed.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,683 @@
+
+    if not check_optimal_termination(res):
+        raise RuntimeError(
+            "Solver failed to terminate with an optimal solution. Please check the solver logs for more details"
```
same here - you can use: `assert_optimal_termination(res)`

ChrisLaliwala replied:
Changed.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,683 @@
+    m.fs.Precipitation.control_volume.properties_in[0].temperature.setub(600)
+
+    ######### S102 Filter Specifications (S/L Filtration following Stage 3)
+    m.fs.S102.split_fraction[0, "liq_outlet", "Liq"].fix(0.9999999)
```
why do we have this? Is there a reason these cannot be 0.99 or 1.0?

ChrisLaliwala replied:
Setting the split fraction to 1.0 leads to flows with a value of 0. This causes convergence issues for the model.

OOAmusat commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,683 @@
+        solver: optimization solver
+        tee: boolean indicator to stream IPOPT solution
+    """
+    # Solve flowsheet
```
why ma57? Is there a reason we do not use the standard ma27?

ChrisLaliwala replied:
Changed to use default ma27. The code is still solving fine.

ksbeattie commented:
@ChrisLaliwala which milestone/deliverable is this tied to?

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_precipitate_liquid_properties.py:
```diff
@@ -0,0 +1,276 @@
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license information.
+#####################################################################################################
+"""
+Initial property package for precipitate.
```
Have deleted this file as there is no need for it. Using file: 'prommis.precipitate.precipitate_liquid_properties'.

ChrisLaliwala commented:
> @ChrisLaliwala which milestone/deliverable is this tied to and when does this need to be merged?

@ksbeattie June is when we promised to have everything uploaded to the GitHub Repo by.

lbianchi-lbl commented:
@ChrisLaliwala FYI, we've moved this to the June release board.

lbianchi-lbl commented:
@ChrisLaliwala the "Check Spelling" failure can be resolved by adding an entry for "equil" in the appropriate section of the `pyproject.toml` file (e.g. under the existing entry for "IDAES"):

https://github.com/prommis/prommis/blob/39d64599dfdd4f4da02946af0b63ce5a4acca067/pyproject.toml#L109-L112

OOAmusat reviewed (approved):
LGTM

bpaul4 commented on pyproject.toml:
```diff
@@ -108,6 +108,8 @@ extend-exclude = [
 [tool.typos.default.extend-words]
 # Ignore IDAES
 IDAES = "IDAES"
+# ignore equil
+equil = "equil"
```
I suspect that this may be a common typo that we still want to catch, if we change this to

```suggestion
# ignore phase_equil
phase_equil = "phase_equil"
```

does it still work?

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_adjustment_rxn_prop_pack.py:
```diff
@@ -0,0 +1,83 @@
+            "heat_of_reaction": constant_dh_rxn,
+            "concentration_form": ConcentrationForm.molarity,
+            "parameter_data": {"dh_rxn_ref": (-1, pyunits.J / pyunits.mol)},
+        },  # Placeholder Value
```
Does this enthalpy value need to be updated?

ChrisLaliwala replied:
This isn't needed, so I removed it. Enthalpies of formation for components were defined in the thermodynamic property package ('cmi_process_prop_pack.py') and are used in the enthalpy balance.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_adjustment_rxn_prop_pack.py:
```diff
@@ -0,0 +1,83 @@
+            "heat_of_reaction": constant_dh_rxn,
+            "concentration_form": ConcentrationForm.molarity,
+            "parameter_data": {"dh_rxn_ref": (-1, pyunits.J / pyunits.mol)},
+        },  # Placeholder Value
```
Same question here.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_dissolution_rxn_prop_pack.py:
```diff
@@ -0,0 +1,83 @@
+            "heat_of_reaction": constant_dh_rxn,
+            "concentration_form": ConcentrationForm.molarity,
+            "parameter_data": {"dh_rxn_ref": (-1, pyunits.J / pyunits.mol)},
+        },  # Placeholder Value
```
Same question here.

ChrisLaliwala replied:
This isn't needed, so I removed it. Enthalpies of formation for components were defined in the thermodynamic property package ('cmi_process_prop_pack.py') and are used in the enthalpy balance.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_dissolution_rxn_prop_pack.py:
```diff
@@ -0,0 +1,83 @@
+            "heat_of_reaction": constant_dh_rxn,
+            "concentration_form": ConcentrationForm.molarity,
+            "parameter_data": {"dh_rxn_ref": (-1, pyunits.J / pyunits.mol)},
+        },  # Placeholder Value
```
Same question here.

ChrisLaliwala replied:
This isn't needed, so I removed it. Enthalpies of formation for components were defined in the thermodynamic property package ('cmi_process_prop_pack.py') and are used in the enthalpy balance.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,670 @@
+from prommis.roasting.ree_oxalate_roaster import REEOxalateRoaster
+
+
+def main():
```
Nice, this looks good. Can you please add a structural diagnostics check just before the solve, and a numerical diagnostics check after the solver results check?

ChrisLaliwala replied:
These have been added.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,670 @@
+    m.fs.Calcination.liquid_in[0].conc_mass_comp["H2O"].fix(1e-5)  # mg/L
+    m.fs.Calcination.frac_comp_recovery.fix(1)
+
+    print(degrees_of_freedom(m))
```
Is this just for debugging, can we take this out?

ChrisLaliwala replied:
This was for debugging. I took out this print statement.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,670 @@
+    )
+
+    results = solver_obj.solve(m, tee=tee)
+    assert_optimal_termination(results)
```
If we have this here, we don't need it on line 72 also.

ChrisLaliwala replied:
This has been removed.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,670 @@
+    )
+
+
+main()
```
Modify so that `main()` runs when the module is executed:

```suggestion
if __name__ == "__main__":
    main()
```

ChrisLaliwala replied:
This has been added.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_precipitation_rxn_prop_pack.py:
```diff
@@ -0,0 +1,83 @@
+            "heat_of_reaction": constant_dh_rxn,
+            "concentration_form": ConcentrationForm.molarity,
+            "parameter_data": {"dh_rxn_ref": (-1, pyunits.J / pyunits.mol)},
+        },  # Placeholder Value
```
Same question here on updating the value.

ChrisLaliwala replied:
This isn't needed, so I removed it. Enthalpies of formation for components were defined in the thermodynamic property package ('cmi_process_prop_pack.py') and are used in the enthalpy balance.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_precipitation_rxn_prop_pack.py:
```diff
@@ -0,0 +1,83 @@
+            "heat_of_reaction": constant_dh_rxn,
+            "concentration_form": ConcentrationForm.molarity,
+            "parameter_data": {"dh_rxn_ref": (-1, pyunits.J / pyunits.mol)},
+        },  # Placeholder Value
```
Same question here.

ChrisLaliwala replied:
This isn't needed, so I removed it. Enthalpies of formation for components were defined in the thermodynamic property package ('cmi_process_prop_pack.py') and are used in the enthalpy balance.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_prop_pack.py:
```diff
@@ -0,0 +1,447 @@
+                "dens_mol_liq_comp_coeff": (
+                    1,
+                    pyunits.kmol * pyunits.m**-3,
+                ),  # PLACEHOLDER
```
Same question for placeholder values in this property package.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/tests/test_cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,815 @@
+from prommis.roasting.ree_oxalate_roaster import REEOxalateRoaster
+
+
+@pytest.fixture(scope="module")
```
I think these should be separate tests:

system_frame() --> m = build()
test_build_flowsheet() checks all objects
test_structural_issues() runs set_operation_conditions() and then diagnostics
test_initialize_system() runs initialize_system()
Don't need test_degrees_of_freedom() since the diagnostics checks for this already.

ChrisLaliwala replied:
I have updated the tests. I have also added a function for scaling the model in 'cmi_process_flowsheet.py' so that is considered here too.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/tests/test_cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,815 @@
+    m = system_frame
+
+    #  FEED (entering Stage 1)
+    assert m.fs.FEED.outlet.flow_mol_phase_comp[0, "Liq", "H2O"].value == pytest.approx(
```
Please check that these all check `pytest.approx([value], rel=1e-4`, if not please specify that explicitly. If the value is zero then you'll have to use `abs-1e-4`.

bpaul4 commented on src/prommis/examples/cmi_process_flowsheet/tests/test_cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,815 @@
+    # Calcinator
+    assert m.fs.Calcination.flow_mol_comp_product[0, "Nd"].value == pytest.approx(
+        0.92499, 1e-4
+    )
```
Add a test to call display_results(), doesn't need to check anything, just call it to confirm that it works.

ChrisLaliwala replied:
This has been added.

bpaul4 reviewed (commented):
Hi @ChrisLaliwala this all looks very good, I have some suggestions for you.

ChrisLaliwala commented:
@bpaul4 I have addressed your comments for this PR. Many of the components disassociate in water so reactions of their ions were considered. This is reflected in the thermo property packages and reaction property packages. Additionally, I added a function for scaling the model to enhance solver performance in the 'cmi_process_flowsheet.py' file. This has led to changes in the 'test_cmi_process_flowsheet.py' file as well.

bpaul4 reviewed (approved):
Everything looks great! Thanks @ChrisLaliwala

MarcusHolly reviewed (changes requested):
A few minor issues to address before merging this

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_prop_pack.py:
```diff
@@ -0,0 +1,487 @@
+from idaes.models.properties.modular_properties.pure.RPP4 import RPP4
+from idaes.models.properties.modular_properties.state_definitions import FpcTP
+
+# all solutes heat capacities approximates as 75000 kJ/kmol/K.
```
This has been changed in the latest commit.

ChrisLaliwala commented on docs/examples/cmi_process.rst:
```diff
@@ -0,0 +1,31 @@
+into iron hydroxide. In stage 3, oxalic acid is added to the solution leading to the formation of soluble iron-ammonium oxalate and insoluble rare earth oxalate. 
+Finally, the rare earth oxalate is filtered from solution and calcined to produce the REO product. 
+
+.. figure:: cmi_process_pfd.png
+    :width: 800
+    :align: center
+
+    Figure 1. CMI Process flowsheet
```
![image](https://github.com/user-attachments/assets/fab1657c-d5d5-4656-a54c-71429abe495f)

MarcusHolly replied:
U should click on the documentation test at the bottom of this PR to view the ReadtheDocs page. I'm still seeing the error, but I'll approve since I'm not sure what the problem is and it's not too big of a deal. @lbianchi-lbl Any idea why the figure is being double-captioned?

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_prop_pack.py:
```diff
@@ -0,0 +1,487 @@
+from pyomo.environ import units as pyunits
+
+# Import IDAES cores
+from idaes.core import (  # Cation,; Anion,; Apparent,
```
This has been removed in the latest commit.

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_adjustment_rxn_prop_pack.py:
```diff
@@ -0,0 +1,74 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
```
This has been fixed.

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_dissolution_rxn_prop_pack.py:
```diff
@@ -0,0 +1,74 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
```
This has been fixed.

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,765 @@
+from pyomo.environ import (
+    ConcreteModel,
```
This has been fixed.

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_precipitation_rxn_prop_pack.py:
```diff
@@ -0,0 +1,74 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
```
This has been fixed.

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_prop_pack.py:
```diff
@@ -0,0 +1,487 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES).
+#
+# Copyright (c) 2018-2023 by the software owners: The Regents of the
+# University of California, through Lawrence Berkeley National Laboratory,
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
```
This has been fixed.

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/tests/test_cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,881 @@
+#####################################################################################################
+# “PrOMMiS” was produced under the DOE Process Optimization and Modeling for Minerals Sustainability
+# (“PrOMMiS”) initiative, and is copyright (c) 2023-2024 by the software owners: The Regents of the
```
This has been fixed.

ChrisLaliwala commented on src/prommis/examples/cmi_process_flowsheet/cmi_process_flowsheet.py:
```diff
@@ -0,0 +1,765 @@
+    """
+    Args:
+        m: pyomo model
+        solver: optimization solver
```
Removed docstring.

ChrisLaliwala commented:
@MarcusHolly I have taken care of your requested changes. Regarding the caption for the figure, it appears correct when I view it GitHub (see screenshot attached as a reply to your original comment).
